### ISSUE TITLE
[Spark Integration] Receiving duplicate OpenLineage events for a Spark Job

### ISSUE THREAD

**Issue by suraj5077** [OpenLineage/OpenLineage#2104]
Context:
Using PySpark (3.4.1) and OL(1.1.0), I tested a simple pipeline which:
1. Reads JSON data into PySpark DataFrame
2. Applies some data transformations
3. Writes transformed data to MySQL database
Observed that we receive 4 events (2 START and 2 COMPLETE) for the same job name. The events are almost identical with a small diff in the facets.
The issue is reproducible and also occurs if writing to a Parquet file instead of SQL.
-------
Slack thread: [https://openlineage.slack.com/archives/C01CK9T7HKR/p1694583867900909](url)
PySpark Code:
```
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, concat_ws, year, month, dayofmonth
from pyspark.sql.types import StructType, StructField, StringType, IntegerType


spark = (SparkSession.builder.master('local')
         .appName('etl-test')
         .config('spark.jars.packages', "io.openlineage:openlineage-spark:1.1.0,mysql:mysql-connector-java:8.0.33")
         .config('spark.extraListeners', 'io.openlineage.spark.agent.OpenLineageSparkListener')
         .config('spark.openlineage.transport.type', 'http')
         .config('spark.openlineage.transport.url', 'http://host.docker.internal:5009/events/spark/')
         .config('spark.openlineage.namespace', 'staging')
         .config('spark.openlineage.transport.auth.type', 'api_key')
         .config('spark.openlineage.transport.auth.apiKey', 'abcdefghijk')
         .getOrCreate())

spark.sparkContext.setLogLevel("INFO")

# Define schema for JSON data
schema = StructType([
  StructField("customer_id", IntegerType(), True),
  StructField("order_id", IntegerType(), True),
  StructField("order_date", StringType(), True),
  StructField("order_amount", StringType(), True),
  StructField("product_name", StringType(), True)
])

# Read JSON data into PySpark DataFrame
df = spark.read.option('multiline', True).json("input.json", schema=schema)

# Apply data transformations
df_transformed = df.select(
  col("customer_id"),
  year(col("order_date")).alias("order_year"),
  month(col("order_date")).alias("order_month"),
  dayofmonth(col("order_date")).alias("order_day"),
  col("order_amount").cast("float").alias("order_amount"),
  concat_ws("-", col("customer_id"), col("order_id")).alias("order_key"),
  col("product_name")
)

# Write transformed data to MySQL database
url = "<url>"
table_name = "orders"
mode = "append"
properties = {
  "driver": "com.mysql.cj.jdbc.Driver",
  "user": "<user>",
  "password": "<pwd>"
}

df_transformed.write.jdbc(url=url, table=table_name, mode=mode, properties=properties)
```

Events emitted:
[spark_events.txt](https://github.com/OpenLineage/OpenLineage/files/12597009/spark_events.txt)

**Comment by arpitsh313** [comment]
Why did you put "staging" there ?
.config('spark.openlineage.namespace', 'staging')

**Comment by suraj5077** [comment]
To define a namespace for all jobs.

**Comment by pawel-big-lebowski** [comment]
Hello @suraj5077, I started working on this immediately because I was aware this is some kind of serious bug introduced recently. However, as I found out, the behaviour like that was present for ~2 years. For some spark SQL jobs, the Openlineage listener gets notified on SQL start/stop and on Job start/stop events. The problem here is that some facets (like iceberg data format versioning) can be only fetched on sql start/stop while some run facets are captured in other case.

We do assume that a single job may emit multiple events and the content of those evens is accumulative which means it is a backend implementation which merges all the events into a single lineage snapshot. However, the problem here is that we should only emit a single `START` and a single `COMPLETE` events, while other shall be sent with `RUNNING` state. This is something I am going to work with my PR.

So, to sum up the story: please do expect still 4 events for this scenario, but there should be only one `start` and one `complete`.

**Comment by suraj5077** [comment]
Hi @pawel-big-lebowski I understand your point. I've also come across cases where the duplicate events have the exact same payload with no difference in facets. I'm assuming there should be some diff in the payload for the backend to merge them. I've also seen a single Job emit 7 events and there is no fixed pattern for this.
I also agree that there should be only a single START and a single COMPLETE event for any Job since this what the OL spec says.

This Job emits 7 events (5 Start and 2 Complete):
```
from pyspark.sql import SparkSession
from pyspark.sql.functions import col
from pyspark.sql.functions import lit
import random

spark = (SparkSession.builder.master('local')
         .appName('spark-pipeline-v1')
         .config('spark.jars.packages', "io.openlineage:openlineage-spark:1.1.0,"
                                        "mysql:mysql-connector-java:8.0.33,"
                                        "net.snowflake:snowflake-jdbc:3.13.14,"
                                        "net.snowflake:spark-snowflake_2.12:2.10.0-spark_3.2")
         .config('spark.extraListeners', 'io.openlineage.spark.agent.OpenLineageSparkListener')
         .config('spark.openlineage.transport.type', 'http')
         .config('spark.openlineage.transport.url', 'http://host.docker.internal:5009')
         .config('spark.openlineage.transport.endpoint', '/events/openlineage/spark/api/v1/lineage')
         .config('spark.openlineage.namespace', 'staging')
         .config('spark.openlineage.transport.auth.type', 'api_key')
         .config('spark.openlineage.transport.auth.apiKey', 'test-key')
         .config('spark.openlineage.parentJobName', 'suraj-test-job')
         .config('spark.openlineage.parentRunId', 'acd-eheh-ththth-wnjwnj')
         .getOrCreate())


mysql_connection_properties = {
    "user": "<user>",
    "password": "<pwd>",
    "driver": "com.mysql.cj.jdbc.Driver",
}

snowflake_options = {
    "sfURL": "<account>.snowflakecomputing.com",
    "sfUser": "<usr>",
    "sfPassword": "<pwd>",
    "sfDatabase": "ANALYTICS",
    "sfWarehouse": "COMPUTE_WH",
    "sfSchema": "PUBLIC",
    "sfRole": "ACCOUNTADMIN",
}

mysql_url = "<mysql-jdbc-url>"

cats_df = spark.read.jdbc(url=mysql_url, table="cats", properties=mysql_connection_properties)
dogs_df = spark.read.jdbc(url=mysql_url, table="dogs", properties=mysql_connection_properties)

cats_df = cats_df.withColumnRenamed("Country", "cat_name")

joined_df = cats_df.join(dogs_df, on="owner", how="inner")

filtered_df = joined_df.filter(col("cat_name") == 'Cookie')

agg_df = filtered_df.groupBy("cat_name").count()

agg_df = agg_df.withColumn("id", lit(random.randint(0, 1000)))

agg_df.write \
    .format("snowflake") \
    .options(**snowflake_options) \
    .option("dbtable", "COUNT_TABLE") \
    .mode("overwrite") \
    .save()

spark.stop()
```

**Comment by suraj5077** [comment]
Since this is a crucial bug, when can we expect a resolution for this?
And thanks for looking into it.

**Comment by suraj5077** [comment]
@pawel-big-lebowski is the above single Job with 7 events an expected behaviour?
